core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1244,6 +1244,61 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When either the inner or outer stream fails, the entire streams fails.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+      Stream.eval(Deferred[F2, Unit]).flatMap { haltOuter =>
+        Stream.eval(async.unboundedQueue[F2, Option[Either[Throwable, Chunk[O2]]]]).flatMap {
+          queue =>
```

Comment:
Yeah, I was trying to match the style of the first 7 lines in `merge`, probably took it too far.

Reply:
sorry, I don't mean these, these are fine. I mean the ones below, the actual streams

Reply:
```scala
 queue.dequeue.unNoneTerminate.rethrow
                     .flatMap(Stream.chunk(_))
                     .concurrently
```

```scala
pulse.dequeue.unNoneTerminate
                       .onFinalize(queue.enqueue1(None))
                       .evalMap {...}.merge
```
```scala
this.attempt
                         .interruptWhen(haltOuter.get.attempt)
                         .onFinalize(pulse.enqueue1(None))
                         .evalMap {
                           case Left(e) =>
                             haltRef.modify(x => None -> x.fold(F2.unit)(_.complete(()))).flatten *>
                               guard.withPermit(queue.enqueue1(Left(e).some))
                           case Right(x) =>
                             Deferred[F2, Unit].flatMap(halt =>
                               haltRef.modify {
                                 case None       => halt.some -> F2.unit
                                 case Some(last) => halt.some -> last.complete(())
                               }.flatten *> pulse.enqueue1((x -> halt).some))
                         }))
```

```scala
case (o, haltInner) =>
                           guard.withPermit(
                             f(o).chunks
                               .evalMap(x => queue.enqueue1(Right(x).some))
                               .interruptWhen(haltInner.get.attempt)
                               .interruptWhen(haltOuter.get.attempt)
                               .compile
                               .drain
                               .handleErrorWith(e =>
                                 haltOuter.complete(()) *> queue.enqueue1(Left(e).some)))
```

at least some of these should be named, ideally you will end up with something like
```scala
foo.concurrently(bar merge baz)
```

Rough guidelines ofc, use your judgement. At the moment it's a bit hard to follow the logic completely (although tbf this is a tricky combinator). Same goes for `queue` and `pulse`, the names don't tell much, is `queue` something like `outQ` (so the final output stream).

Reply:
Done. Yeah, my initial implementation on gitter it way too optimistic, concurrency is hard.

Reply:
Speaking of which, what's the role of the Semaphore here?

Reply:
Basically, the idea is that it prevents the downstream from getting multiple errors as both the inner and outer can terminate the downstream. It seems redundant now because there's a rethrow at downstream and after some testing it seems to work fine without it.